src/python/pants/engine/target.py

```diff
@@ -1343,15 +1343,39 @@ def filespec(self) -> Filespec:
         )
 
 
-@dataclass(frozen=True)
+@frozen_after_init
+@dataclass(unsafe_hash=True)
 class HydrateSourcesRequest:
     field: Sources
+    valid_sources_types: Tuple[Type[Sources], ...]
+
+    def __init__(
+        self, field: Sources, *, valid_sources_types: Iterable[Type[Sources]] = (Sources,)
+    ) -> None:
+        """Convert raw sources globs into an instance of HydratedSources.
+
+        If you only want to convert certain Sources fields, such as only PythonSources, set
```

Comment:
s/convert/handle/

You could probably also the fact that you might get subclasses of your requested types, but that the output type will always be an exact match.